advanced_alchemy/repository/_async.py

```diff
@@ -1016,26 +1016,40 @@ async def upsert_many(
         match_fields = match_fields or self.match_fields
         if isinstance(match_fields, str):
             match_fields = [match_fields]
-        match_filter: list[FilterTypes | ColumnElement[bool]] = [
-            CollectionFilter(
-                field_name=self.id_attribute,
-                values=[getattr(datum, self.id_attribute) for datum in data if datum is not None] if data else None,
-            ),
-        ]
+        if match_fields is None:
+            match_fields = [self.id_attribute]
+        match_filter: list[FilterTypes | ColumnElement[bool]] = []
         if match_fields:
             for field_name in match_fields:
                 field = get_instrumented_attr(self.model_type, field_name)
-                matched_values = [getattr(datum, field_name) for datum in data if datum is not None]
+                matched_values = [
+                    getattr(datum, field_name) for datum in data if getattr(datum, field_name) is not None
+                ]
```

Comment:
```suggestion
                    field_data for datum in data if (field_data := getattr(datum, field_name)) is not None
                ]
```

avoids the duplicated `getattr` call